### DISCUSSION TITLE
MQTT via IOBROKER setzen

### DISCUSSION THREAD

**Discussion by heiner1976** [dewenni/ESP_Buderus_KM271#58] [General]
Hallo liebes Forum,
leider schaffe ich es nicht via IObroker und MQTT Werte zu setzen. 
Über das Webfrontend funktioniert alles perfelt- Ich bekomme auch alle Objekte in IOBroker geschrieben.
Wo muss ich die Werte denn ändern?
Ich habe probiert
Buderus/config/WW_Zirkulation auf 7 zu setzen
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/77c97fc6-2e88-480a-99c9-57a34df10ecc)
leider passiert nichts
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/715dff27-4ebb-4d33-bebd-ef82a999b90a)

Dann dachte ich vielleicht man muss unter Message was einfügen:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/0716b62f-eb42-4750-a778-d68db641b45e)
da passiert aber auch nichts.
Kann mir jemand helfen, was ich falsch mache?
LG
Heiner

Könnte sein, dass dies mit  **"Umschaltung der Betriebsarten per mqtt geht nicht mehr"** vergleichbar ist, da fehlen mir aber die Erfarhungswerte. Ich glaube ich mach es einfach falsch.

**Comment by dewenni** [comment]
Hallo Heiner,

Du kannst nicht einfach die Statuswerte beschreiben. Config und Status sind reine Istwerte zur Anzeige.

Für die Kommandos gibt es extra Befehle und MQTT topics.

Schau mal hier #38

Falls dir das nicht hilft, gib nochmal Bescheid.

Grüße Sven

**Comment by heiner1976** [comment]
Hallo Sven, 
danke schonmal  für die Hilfe. Ich glaube ich bin schon weiter. Ich habe mir mal ein paar mqtt Tutorials angeschaut.
Ich habe jetzt den mqtt Broker bei mir auf dem Rechner und versuche Befehle abzusetzen. 
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/8abc7f30-644f-4a3f-b36d-6197a2ccbbcc)
Ich bekomme nun auch unter Objekten ein Reiter setvalue angezeigt. aber ausgeführt wird nichts.

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/f76d53f3-a428-4919-960f-be7845d47893)

Meine Instanz mqtt.0 ist so eingestellt
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/65397af4-2339-48e5-94da-18d062d6b528)
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/21b57856-6288-4e5d-a64b-6da7642e43a2)

Hast Du vielleicht noch ein Tip für mich?
lg
Heiner

**Comment by dewenni** [comment]
Hallo Heiner,

du kommst der Sache schon näher :-)
Kleiner Hinweis: der MQTT-Explorer ist auch nur ein Client und nicht der Broker. Der Broker läuft bei dir vermutlich im ioBroker.

Die Befehle sind abhängig der eingestellten Sprache im WebUI der Software auch entweder in englisch oder deutsch.
Wenn deine Config und Status Werte über MQTT in deutsch sind, dann müssen auch die Befehle in deutsch gesendet werden.

In deinem Fall müsste das dann so aussehen (bzw. musst du "esp_heizung" bei dir durch "Buderus" ersetzten)
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/a97c2e6b-a48c-4c62-b0c5-cf284feb1531)

Danach taucht sollte das auch als Objekt im ioBroker auftauchen.
Dort kannst du es dann auch beschreiben.

Alternativ kann man die Objekte zum schreiben auch manuell im ioBroker erzeugen.
Ich bin aber zugegeben kein ioBroker Experte.

**Comment by heiner1976** [comment]
Hi Sven,
ich hatte es soweit, das Mqtt - Änderungen nach einem Neustart der Software, übernommen wurden.
Nun ist es leider so, dass sobald ich Mqtt in Einstellungen aktiviert habe, die Karte im Setup -Mode hochfährt. Sobald ich in den Einstellungen mqtt deaktiviert habe , komme ich wieder auf das Webfrontend. Ich habe die Karte sogar mit der 3.2.2 überschrieben, das mit dem Accespoint Mode und Mqtt bleibt aber.  
Das kommt mir komisch vor.
Bei mir auf dem Board steht v0.0.7 vom 2023-09-15.

Hast Du eine Idee was ich noch machen kann? liegt es am vollen Speicher des Boards?
lg
Heiner

**Reply by dewenni** [reply]
nimm diesen Haken hier mal raus:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/46074831/33fda1fd-6105-4a11-b924-f0074f59dde9)

das war hier: https://github.com/dewenni/ESP_Buderus_KM271/discussions/38#discussioncomment-6491537 das Problem

**Comment by heiner1976** [comment]
Danke für den Link, ich hatte dies zwar gelesen, hab aber blöderweise nur nach den "Positionen "der Häckchen geschaut und irgendwie überlesen was vorne dran steht. 
Jetzt geht er nicht mehr in den Setup Mode, danke für die schnelle Hilfe. 
Werde mich nun nochmal ans Werk machen und schauen, warum die Änderungen per MQTT  nicht übernommen werden, muss auch bestimmt irgendeine Einstellung sein.

**Reply by dewenni** [reply]
Wenn der Befehl korrekt übernommen wurde, bekommst du per MQTT auch eine "Message" zurück mit "Befehl xy wurde empfangen!"
Sieht man oben in meinem Screenshot.
Wenn da keine Message zurück kommt, dann stimmt der Befehl nicht.

**Comment by heiner1976** [comment]
Mit diesen Einstellungen habe ich es hin bekommen.
Bei mir war noch dieses Flag falsch:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/4a2c2677-08f0-4336-86ff-c9b14a5da571)
Das muss flase sein, mit Haken konnte ich nichts per mqtt absetzen.

mqtt Adapter Einstellungen in iOBroker:
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/3fdf7b33-3721-4542-aee3-3e9477ed3b3e)

![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/8af3a747-bacc-42c9-9249-9e70945985fa)
![image](https://github.com/dewenni/ESP_Buderus_KM271/assets/50792979/ba21e30e-9db5-4ae3-8f16-e010de89dd3f)
Vielleicht hilft es jemand. Danke Sven für Deine Untestützung und dieses grandiose Projekt!!!!
Gruß
Heiner

**Comment by HMMike** [comment]
Sauber mit dem IOBroker zum Parameterschreiben arbeite ich so:

FUnktion, die direkt via SendTo auf den MQTT Adapter ein publish - clientSend ausführt:

`
function SendToMQTT (topic,message) {

    let aJSON={}

    aJSON.topic= topic

    aJSON.message=message

    aJSON.retain=false 

    setState ('0_userdata.0.L2107.SetPoints.CMD_Counter',getState ('0_userdata.0.L2107.SetPoints.CMD_Counter').val+1,true)    

    log (aJSON)

    sendTo('mqtt.0', 'sendMessage2Client', aJSON);

}`

```
'0_userdata.0.L2107.SetPoints.CMD_Counter ist bei mir ein Zahl Datenpunkt, die die Schreibzugriffe, die von IOBroker kommen, mitzählt.

Für Betriebsart ändern sieht das dann so aus:

// EIne Änderung triggert nur, wenn der Ackmnolegde auf false steht

on ({"id": ['0_userdata.0.L2107.SetPoints.HK1_Betriebsart'], "change": 'ne', "ack": false}, function (obj) {

   // Die Betriebsart Ändern !

   log ("Die Betriebsart wurde geändert");

   SendToMQTT ('L2107/setvalue/hk1_betriebsart',obj.state.val)  

});
```

Gruss Michael

**Comment by dremeier** [comment]
Ich muss mich hier einmal dran hängen, denn auch ich bekomme es nicht hin dem ESP mqtt Daten zu senden.
Nutze SW Version 4.1.0.

Mit dem MQTT Explorer (Client) bin ich zu meinem iobroker MQTT-Broker verbunden und publishe fleißig:
![Screenshot 2024-07-02 022905](https://github.com/dewenni/ESP_Buderus_KM271/assets/8443078/c0156b58-28aa-4451-b6df-8ce71f1d0029)
Im mqtt Broker kommt auch alles an:
![Screenshot 2024-07-02 023157](https://github.com/dewenni/ESP_Buderus_KM271/assets/8443078/05bbf58a-d12b-4aae-becf-c5884f4f6d6c)

Den ESP scheint das alles nicht zu interessieren, weder ist in der "Message" ein Hinweis, weder an anderer Stelle die puplished Daten gespiegelt.

Den Broker habe ich wie oben empfohlen eingestellt. Mein Port ist jedoch nicht 1883 weil anderweitig belegt.
Was mache ich falsch?

Hier mal eine Liste welche der ESP wohl subscribed, bitte korrigieren wenn falsch:


```
Betriebsmodi (Operation Modes)
Buderus-KM271/setvalue/HC1_OPMODE
Buderus-KM271/setvalue/HC2_OPMODE

Reduktionsmodi (Reduction Modes)
Buderus-KM271/setvalue/HC1_REDUCTION_MODE
Buderus-KM271/setvalue/HC2_REDUCTION_MODE

Programme
Buderus-KM271/setvalue/HC1_PROGRAM
Buderus-KM271/setvalue/HC2_PROGRAM

Temperaturschwellwerte (Temperature Thresholds)
Buderus-KM271/setvalue/HC1_FROST_THRESHOLD
Buderus-KM271/setvalue/HC1_SUMMER_THRESHOLD
Buderus-KM271/setvalue/HC1_NIGHT_TEMP
Buderus-KM271/setvalue/HC1_DAY_TEMP
Buderus-KM271/setvalue/HC1_HOLIDAY_TEMP
Buderus-KM271/setvalue/HC1_SWITCH_OFF_THRESHOLD
Buderus-KM271/setvalue/HC1_SWITCH_ON_TEMP
Buderus-KM271/setvalue/HC1_INTERPR
Buderus-KM271/setvalue/HC2_FROST_THRESHOLD
Buderus-KM271/setvalue/HC2_SUMMER_THRESHOLD
Buderus-KM271/setvalue/HC2_NIGHT_TEMP
Buderus-KM271/setvalue/HC2_DAY_TEMP
Buderus-KM271/setvalue/HC2_HOLIDAY_TEMP
Buderus-KM271/setvalue/HC2_SWITCH_OFF_THRESHOLD
Buderus-KM271/setvalue/HC2_SWITCH_ON_TEMP
Buderus-KM271/setvalue/HC2_INTERPR

Ferien (Holiday)
Buderus-KM271/setvalue/HC1_HOLIDAY_DAYS
Buderus-KM271/setvalue/HC2_HOLIDAY_DAYS

Warmwasser (Hot Water)
Buderus-KM271/setvalue/WW_OPMODE
Buderus-KM271/setvalue/WW_TEMP
```

**Reply by dewenni** [reply]
Welche Sprache hast du denn bei dem MQTT Einstellungen eingestellt?

**Reply by dremeier** [reply]
auf "Deutsch"
mit "Englisch" gehts dann auch 
Da muss man erstmal drauf kommen:-)!
Ansonsten: großartige Arbeit von Dir, vielen Dank, echt mega!

**Reply by the78mole** [reply]
Macht es überhaupt Sinn, die MQTT-Topics einzudeutschen? In Excel finde ich die eingedeutschten Formeln auch extrem lästig und schlecht portierbar.

**Reply by dremeier** [reply]
Einheitlich alles auf Englisch finde ich auch Sinnvoller.

**Reply by dremeier** [reply]
Was mir noch fehlt wäre die WW-Aufbereitung EIN/AUS zuschalten bzw. den WW Vorrang. Wahrscheinlich gibt das die Steuerung nicht her?

**Reply by the78mole** [reply]
Warmwasser sollte man ausschalten können... Zumindest geht das in meiner ESPhome Version.
[Token-to-Name Mapping](https://github.com/the78mole/esphome_components/blob/9d74d117d01fa6e8576c2e30f0d812f270c2a6d8/components/km271_wifi/const.py#L14) -->[Token-to-Const Mapping](https://github.com/the78mole/esphome_components/blob/9d74d117d01fa6e8576c2e30f0d812f270c2a6d8/components/km271_wifi/km271_params.h#L368) --> [Const-to-Param Mapping](https://github.com/the78mole/esphome_components/blob/9d74d117d01fa6e8576c2e30f0d812f270c2a6d8/components/km271_wifi/km271_params.h#L105)

**Reply by dremeier** [reply]
@ Sven, welches Setvalue wäre das bei der MQTT Version?
 Also WW-Vorrang = aus oder WW-Aufbereitung = aus

**Reply by the78mole** [reply]
Vielleicht hilft das hier? https://github.com/dewenni/ESP_Buderus_KM271/blob/5759600f0e030e61c299dd15dcc8cc0e70e04b29/include/language.h#L54
oder das hier: https://github.com/dewenni/ESP_Buderus_KM271/blob/5759600f0e030e61c299dd15dcc8cc0e70e04b29/include/language.h#L113

**Reply by dremeier** [reply]
ja, aber nicht als 'setvalue', ich sehe es nur im Bereich 'config' und 'status'. Also kann ich die Parameter nicht setzen?

**Reply by the78mole** [reply]
Sorry, da bin ich dann leider raus... Das muss @dewenni beantworten 🤔

**Reply by dewenni** [reply]
WW_Vorrang und WW_Aufbereitung habe ich nicht als Befehl implementiert. Weiß auch nicht ob das überhaupt geht.
Nur die WW_Betriebsart

**Reply by dewenni** [reply]
Bezüglich deutsch/englisch:
ja, rückblickend würde ich es auch alles in englisch machen.
Ich habe aber leider mit deutschen Topics angefangen und diese bei mir auch so in die Datenbank geschrieben und auch mein Grafana damit aufgebaut. Damit kommt man dann leider in den Zwang das kompatibel zu halten.
Daher gibt es jetzt die Möglichkeit die Sprache einzustellen.
Für Neuinstallationen würde ich bei mqtt auch englisch empfehlen.

**Reply by the78mole** [reply]
WW_Vorrang sollte eigentlich gehen, ist Config Parameter 0x0085
Hier ein paar Schnipsel aus meiner ESPhome component:
```
CFG085    = 0x0085, //: "#0 WW Betriebsart, #3 WW Aufbereitung, #5 WW Zirkulation"
{config_ww_priority_mode, CFG085, true, SensorType::BYTE_AT_OFFSET, 3, "CFG_WW_Vorrangschaltung", ""}
```
Aber tatsächlich weiß ich es im Detail nicht. Da müsste man mal die Protokoll-Spec von Buderus genau anschauen.

Getestet hab ich das aber nie.

**Reply by dewenni** [reply]
Config Parameter 0x0085 ist ja erstmal nur lesend. Das habe ich auch implementiert.
Aber als Kommando habe ich das nicht implementiert.

Der lesende Config Wert für WW_Betriebsart ist z.B. 0x0085  Offset 0
Das zugehörige Kommando ist aber 0x0C0E
